File: src/test/java/tech/sirwellington/alchemy/arguments/assertions/AssertionsTest.java
/*
 * Copyright © 2026. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.arguments.assertions;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import tech.sirwellington.alchemy.arguments.AlchemyAssertion;
import tech.sirwellington.alchemy.arguments.Arguments;
import tech.sirwellington.alchemy.arguments.FailedAssertionException;
import tech.sirwellington.alchemy.generator.StringGenerators;
import tech.sirwellington.alchemy.test.AlchemyTest;
import tech.sirwellington.alchemy.test.generation.GenerateDouble;
import tech.sirwellington.alchemy.test.generation.GenerateInteger;
import tech.sirwellington.alchemy.test.generation.GenerateLong;
import tech.sirwellington.alchemy.test.generation.GenerateString;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.notNullValue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;
import static tech.sirwellington.alchemy.arguments.TestHelpers.assertThrowsFailedAssertion;
import static tech.sirwellington.alchemy.arguments.assertions.Assertions.instanceOf;
import static tech.sirwellington.alchemy.test.ThrowableAssertion.assertThrows;
import static tech.sirwellington.alchemy.test.generation.GenerateInteger.Type.POSITIVE;
import static tech.sirwellington.alchemy.test.generation.GenerateString.Type.ALPHABETIC;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@AlchemyTest
class AssertionsTest {

    @GenerateString(ALPHABETIC)
    private String string;

    @GenerateInteger(POSITIVE)
    private int positiveInt;

    @GenerateLong(GenerateLong.Type.POSITIVE)
    private long positiveLong;

    @GenerateDouble(GenerateDouble.Type.POSITIVE)
    private double positiveDouble;

    @Test
    void testNotNull() {
        var instance = Assertions.notNull();
        assertThat(instance, notNullValue());

        var mock = new Object();
        instance.check(mock);
    }

    @Test
    void testNonNullReference() {
        var instance = Assertions.notNull();
        assertThat(instance, notNullValue());
        assertThrowsFailedAssertion(() -> instance.check(null));
    }

    @Test
    void testNullObject() {
        var instance = Assertions.nullObject();
        instance.check(null);
        assertThrowsFailedAssertion(() -> instance.check(string));
    }

    @Test
    void testSameInstanceAs() {
        var instanceOne = Assertions.sameInstanceAs(null);

        assertThat(instanceOne, notNullValue());
        instanceOne.check(null);

        assertThrowsFailedAssertion(() -> instanceOne.check(new Object()));

        var someObject = new Object();
        var instanceTwo = Assertions.sameInstanceAs(someObject);
        instanceTwo.check(someObject);

        var newObject = new Object();
        assertThrowsFailedAssertion(() -> instanceTwo.check(newObject));
    }

    @Test
    void testInstanceOf() {
        Arguments.checkThat(positiveInt).is(instanceOf(Number.class));
        Arguments.checkThat(positiveInt).is(instanceOf(Integer.class));
        Arguments.checkThat(positiveInt).is(instanceOf(Object.class));
        assertThrowsFailedAssertion(
            () -> Arguments.checkThat(positiveInt).is(instanceOf(String.class))
        );
        assertThrowsFailedAssertion(
            () -> Arguments.checkThat(string).is(instanceOf(Number.class))
        );
    }

    @Test
    void testInstanceOfEdgeCases() {
        var assertion = instanceOf(Number.class);
        assertThrows(() -> assertion.check(null));
    }

    @Test
    void testNot() {
        // Given
        AlchemyAssertion<Object> mock = mock();
        doThrow(FailedAssertionException.class)
            .when(mock)
                .check(any());

        var instance = Assertions.not(mock);

        // When
        instance.check("");
        // Then
        verify(mock).check("");

        // When, no exception
        AlchemyAssertion<Object> successMock = mock();
        // Then, exception
        assertThrowsFailedAssertion(() -> Assertions.not(successMock).check(""));
        verify(successMock).check("");
    }

    @Test
    void testNot_EdgeCases() {
        assertThrows(() -> Assertions.not(null))
            .isIllegalArgumentException();
    }

    @Test
    void testEqualTo() {
        var first = string;
        var second = "";

        do {
            second = StringGenerators.strings().get();
        } while (first.equals(second));

        final var instance = Assertions.equalTo(second);

        instance.check(second); // standard equal
        instance.check(new StringBuilder(second).toString()); // copy equals
        assertThrowsFailedAssertion(() -> instance.check(first)); // not equal ❌
    }
}

File: src/test/java/tech/sirwellington/alchemy/arguments/assertions/AssertionsTest.kt
/*
 * Copyright © 2019. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.arguments.assertions

import com.nhaarman.mockito_kotlin.doNothing
import com.nhaarman.mockito_kotlin.doThrow
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import org.hamcrest.Matchers.notNullValue
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mockito.verifyZeroInteractions
import tech.sirwellington.alchemy.arguments.AlchemyAssertion
import tech.sirwellington.alchemy.arguments.FailedAssertionException
import tech.sirwellington.alchemy.arguments.failedAssertion
import tech.sirwellington.alchemy.generator.StringGenerators.Companion.strings
import tech.sirwellington.alchemy.generator.one
import tech.sirwellington.alchemy.test.junit.ThrowableAssertion.assertThrows
import tech.sirwellington.alchemy.test.junit.runners.*

/**

 * @author SirWellington
 */
@RunWith(AlchemyTestRunner::class)
@Repeat
class AssertionsTest
{

    @GenerateString
    private lateinit var string: String

    @GenerateInteger
    private var positiveInt: Int = 0

    @GenerateLong
    private var positiveLong: Long = 0

    @GenerateDouble
    private var positiveDouble: Double = 0.0

    @Before
    fun setUp()
    {
    }

    @DontRepeat
    @Test
    @Throws(Exception::class)
    fun testNotNull()
    {

        val instance = notNull<Any>()

        assertThat(instance, notNullValue())
        Tests.checkForNullCase(instance)

        val mock = mock<Any>()
        instance.check(mock)
        verifyZeroInteractions(mock)
    }

    @DontRepeat
    @Test
    fun testNonNullReference()
    {
        val instance = nonNullReference<Any>()
        assertThat(instance, notNullValue())

        Tests.checkForNullCase(instance)

        val mock = mock<Any>()
        instance.check(mock)
        verifyZeroInteractions(mock)

        assertThrows { instance.check(null) }.failedAssertion()
    }

    @DontRepeat
    @Test
    fun testNullObject()
    {
        val instance = nullObject<Any>()
        assertThat(instance, notNullValue())

        instance.check(null)

        assertThrows { instance.check(string) }.failedAssertion()
    }

    @Test
    fun testSameInstanceAs()
    {
        val instanceOne = sameInstanceAs<Any?>(null)

        //null is the same instance as null
        assertThat(instanceOne, notNullValue())
        instanceOne.check(null)

        //null is not the same instance as any other non-null object
        assertThrows { instanceOne.check("") }.failedAssertion()

        val someObject = Any()
        val instanceTwo = sameInstanceAs(someObject)
        instanceTwo.check(someObject)

        val differentObject = Any()
        assertThrows { instanceTwo.check(differentObject) }.failedAssertion()
    }

    @Test
    fun testInstanceOf()
    {

        val instance = instanceOf<Any>(Number::class.java)

        instance.check(positiveInt)
        instance.check(positiveLong)
        instance.check(positiveDouble)

        assertThrows { instance.check(string) }.failedAssertion()
    }

    @Test
    fun testInstanceOfEdgeCases()
    {
        val assertion = instanceOf<Any>(Number::class.java)
        assertThrows { assertion.check(null) }.failedAssertion()
    }

    @Test
    fun testNot()
    {
        val assertion = mock<AlchemyAssertion<Any>>()

        doThrow(FailedAssertionException())
                .whenever(assertion)
                .check(ArgumentMatchers.any())

        val instance = not(assertion)

        instance.check("")

        doNothing()
                .whenever(assertion)
                .check(ArgumentMatchers.any())

        assertThrows { instance.check("") }.failedAssertion()
    }

    @Test
    fun testNotEdgeCases()
    {
        assertThrows { not<Any>(null!!) }
    }

    @Test
    fun testEqualTo()
    {
        val first = one(strings())
        var second = ""

        do
        {
            second = one(strings())
        } while (first == second)

        val instance = equalTo(second)

        //Check against self should be ok;
        instance.check(second)
        instance.check("" + second)

        assertThrows { instance.check(first) }.failedAssertion()

    }

}
